Correctly walk import lists in AST visitors

petrochenkov opened the pull request:
Closes https://github.com/rust-lang/rust/issues/28075
Closes https://github.com/rust-lang/rust/issues/28388

r? @eddyb 
cc @brson

brson commented:
Awesome fix - look at all that fallout!

eefriedman commented on src/librustc_front/visit.rs:
```diff
@@ -203,26 +203,31 @@ pub fn walk_item<'v, V: Visitor<'v>>(visitor: &mut V, item: &'v Item) {
+                                    identifier: name,
+                                    parameters: PathParameters::none()
+                                });
+                                visitor.visit_path(unsafe{&*(&path as *const _)}, id);
```
visit_path is defined as `fn visit_path(&mut self, path: &'v Path, _id: NodeId) {`, i.e. the lifetime of the path is tied to the the lifetime of the visitor, not the lifetime of the callback.  Also, the current PR visits both the path's components and the path as a whole.  Also, cloning the prefix is a bunch of extra computation which is unlikely to be useful 

Probably the most straightforward thing to do would be to add a method `visit_path_list_item(&mut self, prefix: &'v Path, item: &'v PathListItem)` to Visitor and let the callee do the right thing.

Related issue: we probably also want a method on Visitor to visit the prefix; it would be useful for fixing the handling of `use std::foobar::{};`.

petrochenkov replied:
Yes, the prefix is cloned and components of the prefix are visited repeatedly. I supposed this couldn't be done optimally without introducing a new `visit_xxx` method, but introducing it seemed.. intrusive and too far away from stability.rs. I'll surely do it if it's acceptable.

I also had an idea to desugar `use a::{b, c};` into `use a::b; use a::b;` during lowering to HIR, it would avoid repeated cloning of the prefix, but checks of the prefix components would still be duplicated.

eddyb replied:
The way the visitor is designed is so that certain visitors can "opt in" to observe the original AST/HIR lifetime and be able to hold onto those references.
Unsafe code here is just wrong.
You have two options: if the HIR map builder doesn't override this particular method, you can drop the lifetime.
Otherwise, you'll have go the explicit route @eefriedman mentioned, with one extra tip: you can always write some sort of iterator for the expansions.

petrochenkov commented:
Updated with accordance to @eefriedman 's advice and rebased.
I'm going to seat on this patch a bit more and fix another bug similar to https://github.com/rust-lang/rust/issues/28075, allowing unstable items outside of braces in imports like
```
use std::rt::{};
```

eddyb commented on src/libsyntax/visit.rs:
```diff
@@ -209,33 +212,21 @@ pub fn walk_item<'v, V: Visitor<'v>>(visitor: &mut V, item: &'v Item) {
                         }
+                    } else {
+                        // FIXME: uncomment this and fix the resulting ICE
+                        // visitor.visit_path(prefix, item.id);
```
Where does the ICE come from?

petrochenkov replied:
Privacy checker. It assumes the prefix is not empty.

bluss commented:
@brson This bug was discovered when we removed a gated function. With this fix, we add new breaking changes by disallowing imports of still present, unused, unstable functionality. Maybe it deserves a crater run? If the fallout is bad, maybe even a warning cycle (even if that sounds extreme).

petrochenkov commented:
>I'm going to seat on this patch a bit more and fix another bug similar to #28075, allowing unstable items outside of braces in imports like `use std::rt::{};`

So, fixing this issue would likely require going to librustc_resolve,  resolving the prefix and assigning `PathResolution` to it without actually importing `std::rt` in the process. I'm not familiar with resolve, so I'm just going to fill a new issue and leave it for later. Update for this PR will follow soon either.

petrochenkov commented:
Done.

bors commented:
:umbrella: The latest upstream changes (presumably #28339) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
Rebased.

eddyb commented:
@brson Can we get a crater run?

alexcrichton commented:
I've scheduled crater runs.

alexcrichton commented:
Looks like this [has some relatively significant breakage](https://gist.github.com/anonymous/75ab63b9789652249ec2).

The cause of all the root regressions appears to be that we have an unstable `std::intrinsics` module, but stable contents (because they're reexported). It was assumed that `use std::intrinsics` would be guarded, but this bug let it slide, so all these use cases got pass the feature gater.

I would personally think, however, that we could submit PRs to update all those crates, rebuild the other non-root-regressions to make sure there's no more fallout, and then land this while publishing information about how to migrate. The breakage here may be largely limited to the `std::intrinsics` module, so it may not be that bad.

thoughts @brson?

petrochenkov commented:
Updated with a fix for https://github.com/rust-lang/rust/issues/28388

bors commented:
:umbrella: The latest upstream changes (presumably #28399) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
Rebased.

brson commented:
Yeah, let's submit PR's downstream. I can look into it tomorrow.

bors commented:
:umbrella: The latest upstream changes (presumably #28349) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
Rebased.

alexcrichton commented:
I have submitted PRs upstream to affected crates:

* [rust-compress](https://github.com/alexcrichton/rust-compress/commit/6d940a392502f39e360bef3dc0610c4d349be9be)
* [tendril](https://github.com/servo/tendril/pull/20)
* [gfx](https://github.com/gfx-rs/gfx/pull/803)

alexcrichton commented on src/test/compile-fail/issue-28388-3.rs:
```diff
@@ -0,0 +1,16 @@
+
+// Prefix in imports with empty braces should be resolved and checked privacy, stability, etc.
+
+use std::rt::{}; //~ ERROR use of unstable library feature 'rt'
```
Can you tweak this test to use an `aux-build` with an unstable module instead? This module in theory may be removed over time.

alexcrichton commented on src/test/compile-fail/issue-28075.rs:
```diff
@@ -0,0 +1,21 @@
+
+#![allow(unused_imports)]
+
+use std::thread::{catch_panic, ScopedKey}; //~ ERROR use of unstable library feature 'catch_panic'
```
Similarly to down below, can you make an auxiliary file for this instead of relying on particular features in the standard library being unstable? (these will change over time)

alexcrichton commented:
Once those PRs have been merged/released I think we'll want to do another crater run here now that more bugs have been fixed.

petrochenkov commented:
@alexcrichton 
Tests updated.

alexcrichton commented:
Ok, a [new crater run](https://gist.github.com/anonymous/4437a360058b7bbdfdf9) indicates that this has 4 root regressions, 2 of which were spurious because of timeouts and 2 of which were because libraries are depending on an older version of gfx which did not get the fix in https://github.com/gfx-rs/gfx/pull/803.

@brson, how do you feel about merging? I'm comfortable with the breakage here and I think we may want to make a post on internals and such to ensure that everyone knows about this, but I think this should be good to go.

brson commented:
@alexcrichton Yes, I feel ok about it. Thanks for doing the downstream work.

alexcrichton commented:
@bors: r+ b44cb01bd73b32ffe046b4dcfa549a954d139ab2

bors commented:
:hourglass: Testing commit b44cb01 with merge e9d2587...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/6522), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/6536), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/6475), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6500), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/6479), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6554), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6549), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/6512), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/1499), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/1498), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/1497), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/1499), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/775), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/1481)

bombless commented:
```rust
mod some_module {
    enum _Enum { A, B, _Dummy }
    pub type Enum = self::_Enum;
    pub use self::_Enum::{A, B};
}
fn to_string(e: some_module::Enum) -> String {
    match e {
        some_module::A => 'A'.to_string(),
        some_module::B => 'B'.to_string(),
        _ => "_not_supported".to_string()
    }
}
```
I guess we might need to apply the privacy rule to `type` as well .
